### ISSUE TITLE
`uv pip compile` Ignores Lower Bound for Transitive Dependency

### ISSUE THREAD

**Issue by vladislavlh** [astral-sh/uv#9322]
Hello `uv` team! 👋

First, I want to express my gratitude for the effort and dedication you’ve put into building and maintaining `uv`. It’s a fantastic tool, and I’ve been enjoying exploring its capabilities.

That said, I recently encountered a small issue while trying to compile dependencies with `uv`. Below are the details:

---

#### **Environment**
- **uv version**: `uv 0.5.3 (56d362208 2024-11-19)`
- **Platform**: `MacOS Sequoia 15.1.1`, 
- **Python version**: `Python 3.12.7`

---

#### **Steps to Reproduce**
Given the following input file `reqs.in`:

```
shap
```
When running the command:
```bash
uv pip compile reqs.in -o reqs.txt --verbose
```

##### **Observed Behaviour**
The resulting `reqs.txt` includes a pinned version of `numba` that does not respect the lower bound specified in the `shap` library's `pyproject.toml` file:
```
numba==0.53.1
```
However, as specified in shap's [pyproject.toml](https://github.com/shap/shap/blob/master/pyproject.toml#L24), the dependency for numba should have a lower bound of `>=0.54`.

From the verbose output, we see:
```
DEBUG Adding transitive dependency for shap==0.46.0: numba*
```
This indicates that `uv` did not pick up the correct lower bound for `numba`. The next steps, however, appear to have executed correctly. For example:
```
DEBUG Adding transitive dependency for numba==0.53.1: numpy>=1.15
```
It found the latest version of `numba` that doesn't have an upper bound of an accepted `numpy==2.1.3`

##### **Expected Behaviour**
The compiled reqs.txt should respect the lower bound for `numba` and pin a version `>=0.54` that is compatible with other dependencies. For instance, `pip-compile` correctly produces an output that includes:
```
numba==0.60.0
numpy==2.0.2
...
```

---

#### **Workaround**
Adding an explicit requirement to `reqs.in` resolves the issue:
```
numba>=0.60.0
```

---

Thank you again for all the incredible work you’re doing with `uv`. Let me know if you need additional details.

**Comment by charliermarsh** [comment]
Yeah, it looks like the last release was in June but that constraint was committed in October.

**Comment by charliermarsh** [comment]
If you install from Git, we should respect it (as expected). Otherwise, you'll need to ask the `shap` team to publish a new release to PyPI.

**Comment by charliermarsh** [comment]
(Happy to answer follow-up questions.)

**Comment by charliermarsh** [comment]
It's a specific issue with recent Numba and NumPy releases. pip-compile generates a different resolution, but both resolutions are technically valid. The main issue tracking that kind of undesirable solution is here: https://github.com/astral-sh/uv/issues/8157. I wrote up an explanation of the specifics of Numba and NumPy here: https://github.com/astral-sh/uv/issues/6281#issuecomment-2316240724.
